File: stack_and_queue/monotonous/queue/template.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2023-01-15 10:59:58
 * @LastEditTime: 2023-01-16 09:48:57
 */
package queue

// 单调队列
type MonotoneQueue struct {
	queue []int
}

func InitQueue() MonotoneQueue {
	return MonotoneQueue{
		queue: []int{},
	}
}

func (mq *MonotoneQueue) Empty() bool {
	return len(mq.queue) == 0
}

func (mq *MonotoneQueue) Back() int {
	return mq.queue[len(mq.queue)-1]
}

func (mq *MonotoneQueue) Font() int {
	return mq.queue[0]
}

func (mq *MonotoneQueue) Push(item int) {
	for !mq.Empty() && mq.Back() < item {
		mq.queue = mq.queue[0 : len(mq.queue)-1]
	}
	mq.queue = append(mq.queue, item)
}

// 这里不能无脑pop，因为item可能在push的时候已经被pop出去了
func (mq *MonotoneQueue) Pop(item int) {
	if !mq.Empty() && mq.Font() == item {
		mq.queue = mq.queue[1:]
	}
}

File: stack_and_queue/monotonous/queue/Template.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2023-01-15 11:00:15
 * @LastEditTime: 2023-01-16 09:46:47
 */
package stack_and_queue.monotonous.queue;

import java.util.Deque;
import java.util.LinkedList;

/**
 * 单调队列的模板
 * 
 * 单调队列的作用是既能够维护队列元素先进先出的顺序，又能够正确维护队列中所有元素的最值
 * 如果只需要维护队列中所有元素的最值，那么优先级队列PriorityQueue就能够很好的完成这项任务
 */

public class Template {
    private Deque<Integer> queue;

    public Template() {
        this.queue = new LinkedList<>();
    }

    // 单调队列push的过程相当于胖子压瘦子
    public void push(int x) {
        while (!queue.isEmpty() && queue.peekLast() < x) {
            // queue.removeLast();
            queue.pollLast();
        }
        queue.addLast(x);
    }

    // 有可能在push的过程中，pop过了
    public void pop(int x) {
        if (queue.peekFirst() == x) {
            // queue.pollFirst();
            queue.pop();
        }
    }

    public int max() {
        return queue.peekFirst();
    }
}
